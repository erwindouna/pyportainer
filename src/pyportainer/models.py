"""Asynchronous Python client for Python Portainer."""

from __future__ import annotations
