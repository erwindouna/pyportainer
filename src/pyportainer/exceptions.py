"""Asynchronous Python client for Python Portainer."""
