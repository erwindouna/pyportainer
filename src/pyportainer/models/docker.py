"""Models for Docker API."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class ImageManifestDescriptorPlatform(DataClassORJSONMixin):
    """Represents the platform information of an image manifest descriptor."""

    architecture: str | None = None
    os: str | None = None
    variant: str | None = None
    os_version: str | None = field(default=None, metadata=field_options(alias="os.version"))
    os_features: list[str] | None = field(default=None, metadata=field_options(alias="os.features"))


@dataclass
class ImageManifestDescriptor(DataClassORJSONMixin):
    """Represents an image manifest descriptor."""

    digest: str | None = None
    size: int | None = None
    urls: list[str] | None = None
    annotations: dict[str, str] | None = None
    data: Any | None = None
    platform: ImageManifestDescriptorPlatform | None = None
    media_type: str | None = field(default=None, metadata=field_options(alias="mediaType"))
    artifact_type: Any | None = field(default=None, metadata=field_options(alias="artifactType"))


@dataclass
class Port(DataClassORJSONMixin):
    """Represents a port mapping for a Docker container."""

    private_port: int | None = field(default=None, metadata=field_options(alias="PrivatePort"))
    public_port: int | None = field(default=None, metadata=field_options(alias="PublicPort"))
    type: str | None = field(default=None, metadata=field_options(alias="Type"))


@dataclass
class HostConfig(DataClassORJSONMixin):
    """Represents the host configuration for a Docker container."""

    annotations: dict[str, str] | None = None
    network_mode: str | None = field(default=None, metadata=field_options(alias="NetworkMode"))


@dataclass
class IPAMConfig(DataClassORJSONMixin):
    """Represents the IP Address Management (IPAM) configuration for a Docker container."""

    ipv4_address: str | None = field(default=None, metadata=field_options(alias="IPv4Address"))
    ipv6_address: str | None = field(default=None, metadata=field_options(alias="IPv6Address"))
    link_local_ips: list[str] | None = field(default=None, metadata=field_options(alias="LinkLocalIPs"))


@dataclass
class Network(DataClassORJSONMixin):
    """Represents the network configuration for a Docker container."""

    endpoint_id: str = field(metadata=field_options(alias="EndpointID"))

    links: list[str] | None = field(default=None, metadata=field_options(alias="Links"))
    aliases: list[str] | None = field(default=None, metadata=field_options(alias="Aliases"))
    gateway: str | None = field(default=None, metadata=field_options(alias="Gateway"))
    ipam_config: IPAMConfig | None = field(default=None, metadata=field_options(alias="IPAMConfig"))
    mac_address: str | None = field(default=None, metadata=field_options(alias="MacAddress"))
    driver_opts: dict[str, str] | None = field(default=None, metadata=field_options(alias="DriverOpts"))
    network_id: str | None = field(default=None, metadata=field_options(alias="NetworkID"))
    ip_address: str | None = field(default=None, metadata=field_options(alias="IPAddress"))
    ip_prefix_len: int | None = field(default=None, metadata=field_options(alias="IPPrefixLen"))
    ipv6_gateway: str | None = field(default=None, metadata=field_options(alias="IPv6Gateway"))
    global_ipv6_address: str | None = field(default=None, metadata=field_options(alias="GlobalIPv6Address"))
    global_ipv6_prefix_len: int | None = field(default=None, metadata=field_options(alias="GlobalIPv6PrefixLen"))
    dns_names: list[str] | None = field(default=None, metadata=field_options(alias="DNSNames"))


@dataclass
class NetworkSettings(DataClassORJSONMixin):
    """Represents the network settings for a Docker container."""

    networks: dict[str, Network] | None = field(default=None, metadata=field_options(alias="Networks"))


@dataclass
class Mount(DataClassORJSONMixin):
    """Represents a mount point for a Docker container."""

    type: str | None = field(default=None, metadata=field_options(alias="Type"))
    name: str | None = field(default=None, metadata=field_options(alias="Name"))
    source: str | None = field(default=None, metadata=field_options(alias="Source"))
    destination: str | None = field(default=None, metadata=field_options(alias="Destination"))
    driver: str | None = field(default=None, metadata=field_options(alias="Driver"))
    mode: str | None = field(default=None, metadata=field_options(alias="Mode"))
    rw: bool | None = field(default=None, metadata=field_options(alias="RW"))
    propagation: str | None = field(default=None, metadata=field_options(alias="Propagation"))


@dataclass
class DockerContainer(DataClassORJSONMixin):
    """Represents a Docker container."""

    id: str = field(metadata=field_options(alias="Id"))
    names: list[str] = field(default_factory=list, metadata=field_options(alias="Names"))

    image: str | None = field(default=None, metadata=field_options(alias="Image"))
    command: str | None = field(default=None, metadata=field_options(alias="Command"))
    created: str | None = field(default=None, metadata=field_options(alias="Created"))
    ports: list[Port] | None = field(default=None, metadata=field_options(alias="Ports"))
    labels: dict[str, str] | None = field(default=None, metadata=field_options(alias="Labels"))
    state: str | None = field(default=None, metadata=field_options(alias="State"))
    status: str | None = field(default=None, metadata=field_options(alias="Status"))
    mounts: list[Mount] | None = field(default=None, metadata=field_options(alias="Mounts"))

    image_id: str | None = field(default=None, metadata=field_options(alias="ImageID"))
    image_manifest_descriptor: ImageManifestDescriptor | None = field(default=None, metadata=field_options(alias="ImageManifestDescriptor"))
    size_rw: str | None = field(default=None, metadata=field_options(alias="SizeRw"))
    size_root_fs: str | None = field(default=None, metadata=field_options(alias="SizeRootFs"))
    host_config: HostConfig | None = field(default=None, metadata=field_options(alias="HostConfig"))
    network_settings: NetworkSettings | None = field(default=None, metadata=field_options(alias="NetworkSettings"))


@dataclass
class PidsStats(DataClassORJSONMixin):
    """Represents PID statistics for a Docker container."""

    current: int | None = None


@dataclass
class NetworkStats(DataClassORJSONMixin):
    """Represents network statistics for a Docker container interface."""

    rx_bytes: int | None = None
    rx_dropped: int | None = None
    rx_errors: int | None = None
    rx_packets: int | None = None
    tx_bytes: int | None = None
    tx_dropped: int | None = None
    tx_errors: int | None = None
    tx_packets: int | None = None


@dataclass
class MemoryStatsDetails(DataClassORJSONMixin):  # pylint: disable=too-many-instance-attributes
    """Represents detailed memory statistics for a Docker container."""

    total_pgmajfault: int = field(default=0)
    cache: int = field(default=0)
    mapped_file: int = field(default=0)
    total_inactive_file: int = field(default=0)
    pgpgout: int = field(default=0)
    rss: int = field(default=0)
    total_mapped_file: int = field(default=0)
    writeback: int = field(default=0)
    unevictable: int = field(default=0)
    pgpgin: int = field(default=0)
    total_unevictable: int = field(default=0)
    pgmajfault: int = field(default=0)
    total_rss: int = field(default=0)
    total_rss_huge: int = field(default=0)
    total_writeback: int = field(default=0)
    total_inactive_anon: int = field(default=0)
    rss_huge: int = field(default=0)
    hierarchical_memory_limit: int = field(default=0)
    total_pgfault: int = field(default=0)
    total_active_file: int = field(default=0)
    active_anon: int = field(default=0)
    total_active_anon: int = field(default=0)
    total_pgpgout: int = field(default=0)
    total_cache: int = field(default=0)
    inactive_anon: int = field(default=0)
    active_file: int = field(default=0)
    pgfault: int = field(default=0)
    inactive_file: int = field(default=0)
    total_pgpgin: int = field(default=0)


@dataclass
class MemoryStats(DataClassORJSONMixin):
    """Represents memory statistics for a Docker container."""

    stats: MemoryStatsDetails = field(default_factory=MemoryStatsDetails)
    max_usage: int = field(default=0)
    usage: int = field(default=0)
    failcnt: int = field(default=0)
    limit: int = field(default=0)


@dataclass
class ThrottlingData(DataClassORJSONMixin):
    """Represents CPU throttling data for a Docker container."""

    periods: int = field(default=0)
    throttled_periods: int = field(default=0)
    throttled_time: int = field(default=0)


@dataclass
class CpuUsage(DataClassORJSONMixin):
    """Represents CPU usage statistics for a Docker container."""

    total_usage: int = field(default=0)
    usage_in_kernelmode: int = field(default=0)
    usage_in_usermode: int = field(default=0)
    percpu_usage: list[int] = field(default_factory=list)


@dataclass
class CpuStats(DataClassORJSONMixin):
    """Represents CPU statistics for a Docker container."""

    cpu_usage: CpuUsage = field(default_factory=CpuUsage)
    system_cpu_usage: int = field(default=0)
    online_cpus: int = field(default=0)
    throttling_data: ThrottlingData = field(default_factory=ThrottlingData)


@dataclass
class DockerContainerStats(DataClassORJSONMixin):
    """Represents Docker container statistics."""

    read: str = field(default="")
    preread: str = field(default="")
    pids_stats: PidsStats = field(default_factory=PidsStats)
    networks: dict[str, NetworkStats] = field(default_factory=dict)
    memory_stats: MemoryStats = field(default_factory=MemoryStats)
    blkio_stats: dict[str, Any] = field(default_factory=dict)
    cpu_stats: CpuStats = field(default_factory=CpuStats)
    precpu_stats: CpuStats | None = field(default_factory=CpuStats)
