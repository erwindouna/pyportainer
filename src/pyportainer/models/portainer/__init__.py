"""Models for Portainer API responses and requests."""
