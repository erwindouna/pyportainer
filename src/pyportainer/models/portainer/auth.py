"""Data models for authentication-related payloads and responses."""

from __future__ import annotations

from mashumaro import DataClassDictMixin
from pydantic import Field


class AuthenticatePayload(DataClassDictMixin):
    """Payload for user authentication."""

    password: str = Field(...)
    username: str = Field(...)


class AuthenticateResponse(DataClassDictMixin):
    """Response containing the JWT token."""

    jwt: str | None = Field(None)


class OauthPayload(DataClassDictMixin):
    """Payload for OAuth authentication."""

    code: str | None = Field(None)


class AuthPayload(DataClassDictMixin):
    """Payload for authentication with optional token."""

    token: str | None = Field(default=None, alias="token")
