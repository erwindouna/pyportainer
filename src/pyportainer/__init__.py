"""Asynchronous Python client for Python Portainer."""

from .pyportainer import
from .exceptions import
from .models import

__all__ = []
