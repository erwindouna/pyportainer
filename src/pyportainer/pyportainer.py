"""Asynchronous Python client for Python Portainer."""

from __future__ import annotations

import asyncio
import json
import socket
from dataclasses import dataclass
from datetime import timedelta
from importlib import metadata
from typing import Any, Self
from urllib.parse import urlparse

from aiohttp import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from pyportainer.exceptions import (
    PortainerAuthenticationError,
    PortainerConnectionError,
    PortainerError,
    PortainerNotFoundError,
    PortainerTimeoutError,
)
from pyportainer.models.docker import DockerContainer, DockerContainerStats, ImageInformation, LocalImageInformation
from pyportainer.models.docker_inspect import DockerInfo, DockerInspect, DockerVersion
from pyportainer.models.portainer import Endpoint

try:
    VERSION = metadata.version(__package__)
except metadata.PackageNotFoundError:  # pragma: no cover
    VERSION = "DEV-0.0.0"  # pylint: disable=invalid-name


@dataclass
class Portainer:
    """Main class for handling connections with the Python Portainer API."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    def __init__(
        self,
        api_url: str,
        api_key: str,
        *,
        request_timeout: float = 10.0,
        session: ClientSession | None = None,
    ) -> None:
        """Initialize the Portainer object.

        Args:
        ----
            api_url: URL of the Portainer API.
            api_key: API key for authentication.
            request_timeout: Timeout for requests (in seconds).
            session: Optional aiohttp session to use.

        """
        self._api_key = api_key
        self._request_timeout = request_timeout
        self._session = session

        parsed_url = urlparse(api_url)
        self._api_host = parsed_url.hostname or "localhost"
        self._api_scheme = parsed_url.scheme or "http"
        self._api_port = parsed_url.port or 9000

    # pylint: disable=too-many-arguments, too-many-locals
    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
        json_body: dict[str, Any] | None = None,
        timeout: float | None = None,
        parse: bool = True,
    ) -> Any:
        """Handle a request to the Python Portainer API.

        Args:
        ----
            uri: Request URI, without '/api/', for example, 'status'.
            method: HTTP method to use.
            params: Extra options to improve or limit the response.
            timeout: Timeout for the request (in seconds).
            parse: Whether to parse the response as JSON.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the Python Portainer API.

        Raises:
        ------
            Python PortainerAuthenticationError: If the API key is invalid.

        """
        url = URL.build(
            scheme=self._api_scheme,
            host=self._api_host,
            port=self._api_port,
            path="/api/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json, text/plain",
            "User-Agent": f"PythonPortainer/{VERSION}",
            "X-API-Key": self._api_key,
        }

        if self._session is None:
            self._session = ClientSession()
            self._close_session = True

        # Only override timeout if a specific value is provided, else use default
        if timeout is None:
            timeout = self._request_timeout

        try:
            async with asyncio.timeout(timeout):
                response = await self._session.request(
                    method,
                    url,
                    headers=headers,
                    params=params,
                    json=json_body,
                )
                response.raise_for_status()
        except TimeoutError as err:
            msg = f"Timeout error while accessing {method} {url}: {err}"
            raise PortainerTimeoutError(msg) from err
        except ClientResponseError as err:
            if err.status == 401:
                msg = f"Authentication failed for {method} {url}: Invalid API key"
                raise PortainerAuthenticationError(msg) from err
            if err.status == 404:
                msg = f"Resource not found at {method} {url}: {err}"
                raise PortainerNotFoundError(msg) from err
            msg = f"Connection error for {method} {url}: {err}"
            raise PortainerConnectionError(msg) from err
        except (ClientError, socket.gaierror) as err:
            msg = f"Unexpected error during {method} {url}: {err}"
            raise PortainerConnectionError(msg) from err

        if response.status in (204, 304):
            return None

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Portainer API"
            raise PortainerError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        # Read events instead. Ideal for getting image pull progress
        events: list[Any] = []
        if not parse:
            async for chunk in response.content:
                for line in chunk.splitlines():
                    stripped_line = line.strip()
                    if not stripped_line:
                        continue
                    events.append(json.loads(stripped_line))
            return events

        return await response.json()

    async def get_endpoints(self) -> list[Endpoint]:
        """Get the list of endpoints from the Portainer API.

        Returns
        -------
            A list of Endpoint objects.

        """
        endpoints = await self._request("endpoints")

        return [Endpoint.from_dict(endpoint) for endpoint in endpoints]

    async def get_containers(self, endpoint_id: int) -> list[DockerContainer]:
        """Get the list of containers from the Portainer API.

        Args:
        ----
            endpoint_id: The ID of the endpoint to get containers from.
            all: If True, include all containers. If False, only running containers.

        Returns:
        -------
            A list of containers.

        """
        containers = await self._request(f"endpoints/{endpoint_id}/docker/containers/json?all=1")

        return [DockerContainer.from_dict(container) for container in containers]

    async def start_container(self, endpoint_id: int, container_id: str) -> Any:
        """Start a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to start.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/start",
            method="POST",
        )

    async def stop_container(self, endpoint_id: int, container_id: str) -> Any:
        """Stop a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to stop.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/stop",
            method="POST",
        )

    async def restart_container(self, endpoint_id: int, container_id: str) -> Any:
        """Restart a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to restart.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/restart",
            method="POST",
        )

    async def pause_container(self, endpoint_id: int, container_id: str) -> Any:
        """Pause a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to pause.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/pause",
            method="POST",
        )

    async def unpause_container(self, endpoint_id: int, container_id: str) -> Any:
        """Unpause a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to unpause.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/unpause",
            method="POST",
        )

    async def kill_container(self, endpoint_id: int, container_id: str) -> Any:
        """Kill a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to kill.

        """
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/kill",
            method="POST",
        )

    async def delete_container(self, endpoint_id: int, container_id: str, *, force: bool = False) -> Any:
        """Delete a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to delete.
            force: If True, force delete the container.

        """
        params = {"force": str(force).lower()}
        return await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}",
            method="DELETE",
            params=params,
        )

    async def inspect_container(self, endpoint_id: int, container_id: str, *, raw: bool = False) -> DockerInspect | Any:
        """Inspect a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to inspect.
            raw: If True, return the raw JSON response. If False, return a DockerInspect object.

        Returns:
        -------
            A DockerContainer object with the inspected data.

        """
        container = await self._request(f"endpoints/{endpoint_id}/docker/containers/{container_id}/json")

        if raw:
            return container
        return DockerInspect.from_dict(container)

    async def docker_version(self, endpoint_id: int) -> DockerVersion:
        """Get the Docker version on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.

        Returns:
        -------
            A DockerVersion object with the Docker version data.

        """
        version = await self._request(f"endpoints/{endpoint_id}/docker/version")

        return DockerVersion.from_dict(version)

    async def docker_info(self, endpoint_id: int) -> DockerInfo:
        """Get the Docker info on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.

        Returns:
        -------
            A DockerInfo object with the Docker info data.

        """
        info = await self._request(f"endpoints/{endpoint_id}/docker/info")

        return DockerInfo.from_dict(info)

    async def container_stats(
        self,
        endpoint_id: int,
        container_id: str,
        *,
        stream: bool = False,
        one_shot: bool = True,
    ) -> Any:
        """Get the stats of a container on the specified endpoint.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to get stats from.
            stream: If True, stream the stats. If False, get a single snapshot.
            one_shot: If True, get a single snapshot. If False, stream the stats.

        Returns:
        -------
            The stats of the container.

        """
        params = {"stream": str(stream).lower(), "one-shot": str(one_shot).lower()}
        stats = await self._request(
            f"endpoints/{endpoint_id}/docker/containers/{container_id}/stats",
            params=params,
        )

        return DockerContainerStats.from_dict(stats)

    async def get_image_information(self, endpoint_id: int, image_id: str) -> ImageInformation:
        """Get information about a Docker image.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            image_id: The ID of the image to get information about.

        Returns:
        -------
            An ImageInformation object with the image data.

        """
        image = await self._request(f"endpoints/{endpoint_id}/docker/distribution/{image_id}/json")

        return ImageInformation.from_dict(image)

    async def get_image(self, endpoint_id: int, image_id: str) -> LocalImageInformation:
        """Get information about a Docker image.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            image_id: The ID of the image to get information about.

        Returns:
        -------
            A LocalImageInformation object with the image data.

        """
        image = await self._request(f"endpoints/{endpoint_id}/docker/images/{image_id}/json")

        return LocalImageInformation.from_dict(image)

    async def image_recreate(self, endpoint_id: int, image_id: str, timeout: timedelta = timedelta(minutes=5)) -> Any:
        """Recreate a Docker image.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            image_id: The ID of the image to recreate.
            timeout: Timeout for the image recreation process. Defaults to 5 minutes.

        Returns:
        -------
            An ImageInformation object with the recreated image data.

        """
        params = {"fromImage": image_id}
        return await self._request(
            uri=f"endpoints/{endpoint_id}/docker/images/create?fromImage={image_id}",
            timeout=timeout.total_seconds(),
            method="POST",
            params=params,
            parse=False,
        )

    async def container_recreate_helper(self, endpoint_id: int, container_id: str, image: str, timeout: timedelta = timedelta(minutes=5)) -> Any:
        """Recreate a Docker container service.

        This helper runs through the Portainer API and recreates the specified container.
        It first inspects the container to get its configuration, then creates a new container
        with the same configuration.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            container_id: The ID of the container to recreate.
            image: The tag of the image to use for the new container.

        Returns:
        -------
            The response from the Portainer API.

        """
        container_inspect = await self.inspect_container(
            endpoint_id=endpoint_id,
            container_id=container_id,
            raw=True,
        )

        if not isinstance(container_inspect, dict):
            msg = "Failed to inspect container for recreation."
            raise PortainerError(msg)

        await self.image_recreate(
            endpoint_id=endpoint_id,
            image_id=image,
            timeout=timeout,
        )

        await self.stop_container(
            endpoint_id=endpoint_id,
            container_id=container_id,
        )

        await self.delete_container(
            endpoint_id=endpoint_id,
            container_id=container_id,
            force=True,
        )

        create_body = {
            **container_inspect["Config"],
            "Image": image,
            "HostConfig": container_inspect["HostConfig"],
            "Config": container_inspect["Config"],
        }

        created = await self.container_create(
            endpoint_id=endpoint_id,
            name=container_inspect.get("Name", "").lstrip("/"),
            image=image,
            config=create_body,
        )

        # This is optional; reattach networks the same way as the original container
        # I have to test this, probablt need some friendly users...
        networks = (container_inspect["NetworkSettings"] or {}).get("Networks") or {}
        for net_name in networks:
            network_name = (container_inspect["HostConfig"] or {}).get("NetworkMode") or ""
            if network_name in {"host", "none"} or network_name.startswith("container:"):
                continue

            await self._request(
                f"endpoints/{endpoint_id}/docker/networks/{net_name}/connect",
                method="POST",
                json_body={"Container": created.id},
            )

        await self.start_container(
            endpoint_id=endpoint_id,
            container_id=created.id,
        )

        return created

    async def container_create(self, endpoint_id: int, name: str, image: str, config: dict[str, Any]) -> DockerContainer:
        """Create a Docker container.

        Args:
        ----
            endpoint_id: The ID of the endpoint.
            name: The name of the container to create.

        Returns:
        -------
            A DockerContainer object with the created container data.

        """
        params = {"name": name}
        json_body = {"Image": image}
        json_body.update(config)
        container = await self._request(
            uri=f"endpoints/{endpoint_id}/docker/containers/create",
            method="POST",
            params=params,
            json_body=json_body,
        )

        return DockerContainer.from_dict(container)

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Portainer object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
