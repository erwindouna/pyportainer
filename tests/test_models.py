"""Test the models of the Portainer API."""

from __future__ import annotations

from typing import TYPE_CHECKING

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from tests import load_fixtures

if TYPE_CHECKING:
    from pyportainer import Portainer
    from pyportainer.models.portainer import Endpoint


async def test_portainer_endpoints(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    portainer_client: Portainer,
) -> None:
    """Test the Portainer endpoints."""
    aresponses.add(
        "localhost:9000",
        "/api/endpoints",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("endpoints.json"),
        ),
    )

    endpoints: list[Endpoint] = await portainer_client.get_endpoints()
    assert endpoints == snapshot


async def test_portainer_containers(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    portainer_client: Portainer,
) -> None:
    """Test the Portainer containers."""
    aresponses.add(
        "localhost:9000",
        "/api/endpoints/1/docker/containers/json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("containers.json"),
        ),
    )

    containers = await portainer_client.get_containers(1)
    assert containers == snapshot
