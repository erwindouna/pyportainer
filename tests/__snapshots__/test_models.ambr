# serializer version: 1
# name: test_portainer_containers
  list([
    DockerContainer(names=None, image=None, command=None, created=None, ports=None, labels=None, state=None, status=None, mounts=None, id='aa86eacfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf', image_id='sha256:72297848456d5d37d1262630108ab308d3e9ec7ed1c3286a32fe09856619a782', image_manifest_descriptor=ImageManifestDescriptor(digest='sha256:c0537ff6a5218ef531ece93d4984efc99bbf3f7497c0a7726c88e2bb7584dc96', size=424, urls=['http://example.com'], annotations={'com.docker.official-images.bashbrew.arch': 'amd64', 'org.opencontainers.image.base.digest': 'sha256:0d0ef5c914d3ea700147da1bd050c59edb8bb12ca312f3800b29d7c8087eabd8', 'org.opencontainers.image.base.name': 'scratch', 'org.opencontainers.image.created': '2025-01-27T00:00:00Z', 'org.opencontainers.image.revision': '9fabb4bad5138435b01857e2fe9363e2dc5f6a79', 'org.opencontainers.image.source': 'https://git.launchpad.net/cloud-images/+oci/ubuntu-base', 'org.opencontainers.image.url': 'https://hub.docker.com/_/ubuntu', 'org.opencontainers.image.version': '24.04'}, data=None, platform=ImageManifestDescriptorPlatform(architecture='arm', os='windows', variant='v7', os_version='10.0.19041.1165', os_features=['win32k']), media_type='application/vnd.oci.image.manifest.v1+json', artifact_type=None), size_rw='122880', size_root_fs='1653948416', host_config=HostConfig(annotations=None, network_mode='mynetwork'), network_settings=NetworkSettings(networks=None)),
  ])
# ---
# name: test_portainer_endpoints
  list([
    Endpoint(amt_device_guid='4c4c4544-004b-3910-8037-b6c04f504633', authorized_teams=[1], authorized_users=[1], azure_credentials={'ApplicationID': 'eag7cdo9-o09l-9i83-9dO9-f0b23oe78db4', 'AuthenticationKey': 'cOrXoK/1D35w8YQ8nH1/8ZGwzz45JIYD5jxHKXEQknk=', 'TenantID': '34ddc78d-4fel-2358-8cc1-df84c8o839f5'}, compose_syntax_max_version='3.8', container_engine='docker', edge_checkin_interval=5, edge_id='string', edge_key='string', enable_gpu_management=True, gpus=[{'name': 'name', 'value': 'value'}], group_id=1, heartbeat=True, id=1, is_edge_device=True, kubernetes={'Configuration': {'AllowNoneIngressClass': True, 'EnableResourceOverCommit': True, 'IngressAvailabilityPerNamespace': True, 'IngressClasses': [{'Blocked': True, 'BlockedNamespaces': ['string'], 'Name': 'string', 'Type': 'string'}], 'ResourceOverCommitPercentage': 0, 'RestrictDefaultNamespace': True, 'StorageClasses': [{'AccessModes': ['string'], 'AllowVolumeExpansion': True, 'Name': 'string', 'Provisioner': 'string'}], 'UseLoadBalancer': True, 'UseServerMetrics': True}, 'Flags': {'IsServerIngressClassDetected': True, 'IsServerMetricsDetected': True, 'IsServerStorageDetected': True}, 'Snapshots': [{'DiagnosticsData': {'DNS': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}, 'Log': 'string', 'Proxy': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}, 'Telnet': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}}, 'KubernetesVersion': 'string', 'NodeCount': 0, 'Time': 0, 'TotalCPU': 0, 'TotalMemory': 0}]}, name='my-environment', public_url='docker.mydomain.tld:2375', snapshots=[{'ContainerCount': 0, 'DiagnosticsData': {'DNS': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}, 'Log': 'string', 'Proxy': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}, 'Telnet': {'additionalProp1': 'string', 'additionalProp2': 'string', 'additionalProp3': 'string'}}, 'DockerSnapshotRaw': {}, 'DockerVersion': 'string', 'GpuUseAll': True, 'GpuUseList': ['string'], 'HealthyContainerCount': 0, 'ImageCount': 0, 'IsPodman': True, 'NodeCount': 0, 'RunningContainerCount': 0, 'ServiceCount': 0, 'StackCount': 0, 'StoppedContainerCount': 0, 'Swarm': True, 'Time': 0, 'TotalCPU': 0, 'TotalMemory': 0, 'UnhealthyContainerCount': 0, 'VolumeCount': 0}], status=1, tls=True, tls_ca_cert='string', tls_cert='string', tls_key='string', tag_ids=[1], tags=['string'], team_access_policies={'additionalProp1': {'RoleId': 1}, 'additionalProp2': {'RoleId': 1}, 'additionalProp3': {'RoleId': 1}}, type=1, url='docker.mydomain.tld:2375', user_access_policies={'additionalProp1': {'RoleId': 1}, 'additionalProp2': {'RoleId': 1}, 'additionalProp3': {'RoleId': 1}}, user_trusted=True, tls_config=TLSConfig(tls=True, tls_ca_cert='/data/tls/ca.pem', tls_cert='/data/tls/cert.pem', tls_key='/data/tls/key.pem', tls_skip_verify=False), agent=Agent(version='1.0.0'), edge=Edge(command_interval=60, ping_interval=60, snapshot_interval=60, async_mode=True), last_check_in_date=0, query_date=0, security_settings=SecuritySettings(allow_bind_mounts_for_regular_users=False, allow_container_capabilities_for_regular_users=True, allow_device_mapping_for_regular_users=True, allow_host_namespace_for_regular_users=True, allow_privileged_mode_for_regular_users=False, allow_stack_management_for_regular_users=True, allow_sysctl_setting_for_regular_users=True, allow_volume_browser_for_regular_users=True, enable_host_management_features=True)),
  ])
# ---
